book/src/rfc/rfc-0-generic-parameters.md

```diff
@@ -519,6 +519,35 @@ fn main() {
 
 The current `compute_type` function can remain as it is. It stores the computed concrete types, such as `Array` or `Custom` as shown in the examples above, instead of the generic types. It creates a mapping between the expression node from the AST and the type `TypeKind`.
 
+### Function Call Instantiation
+The functions are defined as `FunctionDef`, which is an AST containing the signature and the body of the function. The body is a vector of statements, each of which is a tree of expression nodes. It is fine to have different function calls pointing to these functions' original AST, when the content of these functions doesn't change, and so are the expression nodes.
+
+However, when a function takes generic arguments, the actual arguments can result in different expression nodes. For example:
+
+```rust
+fn last(arr: [Field; N]) -> Field {
+    return arr[N - 1];
+}
+
+fn main() {
+    let arr1 = [1, 2, 3, 4, 5];
+    let arr2 = [6, 7, 8, 9];
+
+    let last1 = last(arr1); // instantiate `last` function with N = 5
+    let last2 = last(arr2); // instantiate `last` function with N = 4
+}
+```
+
+The monomorphized body of the function call for `last(arr1)` is `return arr[5 - 1]`, while the one for `last(arr2)` is `return arr[4 - 1]`. Therefore, we can't have a single expression node representing both `arr[5 - 1]` and `arr[4 - 1]` expression nodes. These functions should be instantiated with new ASTs, which are monomorphized from the original ASTs. They will be regenerated with the generic parameters being resolved with concrete values. 
+
+To ensure no conflicts in the node IDs being regenerated for these instantiated functions, the AST for the main function as an entry point will be regenerated. The monomorphized AST preserves the span information to point to the noname source code for the existing debugging feature.
+
+Same as before, these instantiated functions can be pointed by the expression nodes `ExprKind::FnCall`. With the support of generic parameters, we need to change the way of loading the function AST, as the current fully qualified name pattern doesn't contain the information to differentiate the instantiated functions with different generic values.
+
+The new string pattern to store the monomorphized function AST can be:
+`fn_full_qualified_name#generic1=value1#generic2=value2`
```

Comment:
👍

Reply:
so can you add a blurb right here on how the monormophization really works? My idea now is that monomorphization walks through the AST and:

1. propagate types in the same way the type checker was doing but replaces any generic type with a concrete type + folds constants + performs some type checking when it converts a generic type into a concrete type
2. whenever it encounters a generic function call it copies its AST into a new function `new_name = monomorphize_name(function_name, generics_and_values)` stored in the AST, then walks through its AST (modification happening due to 1 would apply on the AST of `new_name`. 
3. at the end it deletes every generic functions from the AST

the important thing is that in 1. when a generic type is converted into a concrete type, if we are inside of a generic function then we need to assign the node a new node ID

Reply:
so just to drive the point I was making last time, I think with that change we could have that the monomorphization pass only mutates the TAST, making it a simple wrapper at the end: `Mast<B: Backend>(TypeChecker<B>)`

It would only ever mutate the TAST to:

1. add new functions in `functions: HashMap<FullyQualified, FnInfo<B>>,`
2. convert generic types into concrete types in `node_types: HashMap<usize, TyKind>,`
3. assign a new node ID to instantiated/monomorphized functions nodes that are generic types and register them in `node_types: HashMap<usize, TyKind>,`

Reply:
Added.

We also have to modify the `FnCall` node, otherwise it won't point to the monomorphized function ast.